src/ui/popupmenu/completion_item_widget.rs

```diff
@@ -144,6 +165,12 @@ fn get_icon_name_for_kind(kind: &str, color: &Color, size: f64) -> String {
         "snippet" => icon!("../../../assets/icons/file-text.svg", color, size),
         "folder" => icon!("../../../assets/icons/folder.svg", color, size),
 
-        _ => icon!("../../../assets/icons/help-circle.svg", color, size),
+        _ => {
+            if know_items_kinds {
+                icon!("../../../assets/icons/help-circle.svg", color, size)
+            } else {
+                String::from("")
```

Comment:
What do you think of always returning `String::from("")` or something that's easy to identify as not matched (e.g. `None`).

And then on the place that assigns icons, doing something like this (I am not too familiar with rust, pseudocode):

```
// Get all are not matched ahead of time
let any_kind_matches = all(state.items_to_load, (i) => Some(get_icon_pixbuf(&i.kind, .., ..))

// ...
// specific item
if any_kind_matches {
  prev.kind.set_from_pixbuf(&buf.unwrap());
}
```

This way we void passing too much state for this, just getting the variable once. I hope I was able to communicate clearly, let me know if that's not the case.

Reply:
So you're saying that we should do the checks for which icon should be used outside of the `get_icon_name_for_kind` and `get_icon_pixbuf` functions?

If that's what you mean (correct me if I'm wrong), then I totally agree. We could instead create a function that takes all of the popupmenu's CompletionItems as a parameter (in addition to all of the parameters of `get_icon_pixbuf`), do the needed checks inside of that function to determine whether or not a help-circle (or another, TBD icon) should be used, and then return the necessary pixbuf. Thus, we don't needlessly pass around too much state.

Is that what you were thinking, or did I misunderstand?

Reply:
> @badosu So you're saying that we should do the checks for which icon should be used outside of the get_icon_name_for_kind and get_icon_pixbuf functions?

No, what I am trying to convey is to separate the case where no match was found from the `get_icon_name_for_kind`. So if it's not found we can just make a check to see if it was not found for all items. If that's the case then we can decide if we set the question mark icon or not (whether all items are matched or not).

The idea here is that we shouldn't need to pass the information if all items are matched (`know_items_kinds`) inside any other scope than the one where we have all items and the one that sets the pixbuf.


I tried coding this, but was having some issues with the `icon!` macro that prevented me from proceeding forward.

Reply:
Ok, that makes sense, and I like it. In order to implement it, however, I feel as though `get_icon_pixbuf` gets in the way. What do you think about making `get_icon_name_for_kind` public and then calling it directly? If `get_icon_name_for_kind` returned "None" or an empty string we would then check the strings returned for all of the other items to see if they are "None" or empty likewise. Afterwards we call `get_icon_pixbuf` with the necessary `contents` parameter (help-circle icon if we need to use an icon). We could probably not call `get_icon_pixbuf` at all in `CompletionItemWidgetWrap::create()` and use an empty image instead if we didn't need the help-circle.

Reply:
I am still a beginner on Rust, so take my suggestions with a grain of salt :-).

Reply:
It’s all good (I’m still learning the lang. too 😉), and your suggestions have been great. So you want me to push the change I stated above? (Also, check out Rust’s “book” if you haven’t already and get the chance, great doc for Rust.)

Reply:
In order to check the other items' kinds in the `CompletionItemWidgetWrap::create()` function, I would need to pass it the other CompletionItems as a parameter (instead of `know_items_kinds`). Sound good?